### ISSUE TITLE
OutOfMemoryError when IO.uncancelable is used in recursive function

### ISSUE THREAD

**Issue by tomohavvk** [typelevel/cats-effect#4043]
Version of **cats-effect**: 3.5.4

While going through the tutorial [Producer consumer with taking care of cancellation](https://typelevel.org/cats-effect/docs/tutorial#taking-care-of-cancelation) I noticed that the code provided in the example terminates with a `java.lang.OutOfMemoryError`. The timing of the error's occurrence may vary depending on the `Xmx` **VM option**, but the error still occurs nonetheless. 
I'm trying to understand whether this is the expected behavior and whether I should avoid using `IO.uncancelable` in recursive code, or if it's a BUG?

If this is the expected behaviour, it seems strange that the **Typelevel** website showcases an example implementation of a producer and consumer that eventually leads to an out-of-memory error due to memory overflow. 
IMHO at least the behaviour should be mentioned in the tutorial.

Below is an example code that includes two programs. **badProgram** leads to an `java.lang.OutOfMemoryError` when using `IO.uncancelable` in recursive function, while the **goodProgram** works indefinitely without a loss in performance until manually terminated.

Code example - https://scastie.scala-lang.org/bjx9e0WsRq20XTxUprnCsg

```scala
import cats.effect._
import java.time.{LocalDateTime, ZoneOffset}

object OOMBug extends IOApp {

  override def run(args: List[String]): IO[ExitCode] =
    for {
      //      _ <- goodProgram(0)
      _ <- badProgram(0)
    } yield ExitCode.Success

  private val badProgram: Long => IO[Long] = counter =>
    IO.uncancelable { poll =>
      printIfNeed(counter) >> poll(badProgram(counter + 1)).onCancel(IO.unit)
    }

  private val goodProgram: Long => IO[Long] = counter =>
    pollImitation { poll =>
      printIfNeed(counter) >> poll(goodProgram(counter + 1))
    }

  private val printIfNeed: Long => IO[Unit] = counter =>
    IO.whenA(counter % 100000 == 0) {
      IO.delay(LocalDateTime.now.toInstant(ZoneOffset.UTC).toEpochMilli)
        .flatMap { millis =>
          IO.println(s"Time: $millis. Counter: $counter")
        }
    }

  private def pollImitation[A](body: Poll[IO] => IO[A]): IO[A] =
    body {
      new Poll[IO] {
        def apply[B](ioa: IO[B]): IO[B] = ioa
      }
    }
}
```

Error output:
```log
Time: 1710428772216. Counter: 33400000
Time: 1710428772267. Counter: 33500000
2024-03-14T15:06:13.650Z [WARNING] Your app's responsiveness to a new asynchronous event (such as a new connection, an upstream response, or a timer) was in excess of 100 milliseconds. Your CPU is probably starving. Consider increasing the granularity of your delays or adding more cedes. This may also be a sign that you are unintentionally running blocking I/O operations (such as File or InetAddress) without the blocking combinator.
java.lang.OutOfMemoryError: Java heap space
	at cats.effect.ByteStack$.growIfNeeded(ByteStack.scala:48)
	at cats.effect.ByteStack$.push(ByteStack.scala:56)
	at cats.effect.IOFiber.runLoop(IOFiber.scala:582)
	at cats.effect.IOFiber.autoCedeR(IOFiber.scala:1423)
	at cats.effect.IOFiber.run(IOFiber.scala:119)
	at cats.effect.unsafe.WorkerThread.run(WorkerThread.scala:743)
```

**Comment by djspiewak** [comment]
This is very subtle! So a couple things are going on here…

First, the example on the linked documentation page doesn't actually crash with OOM *if* you use the [better-monadic-for plugin](https://github.com/oleg-py/better-monadic-for). We should probably make that more clear. The reason is the recursive `producer`/`consumer` invocations are the last step of a `for`-comprehension. This is totally valid and fine, but without the plugin it results in desugared code which will eventually go OOM.

Second, your code obviously *also* goes OOM, but for a very different reason! The essence of the issue here is that the code is not tail-recursive within `IO`. This is a slightly different definition of tail-recursion than what is typical within Scala, but it's the same conceptual idea. In this case, the work done after the recursive call returns is the resolution of the `poll`, the `onCancel`, and the closing of the `uncancelable`. That work *fundamentally* accumulates with each recursive stage, so it's always going to exhaust resources. With direct calls, this would manifest as a `StackOverflowError`. With `IO` it manifests (eventually) as an `OutOfMemoryException`.

So in a sense, this is behaving as expected. In general, you don't want to (and don't have to!) use `uncancelable` in this type of *unbounded* recursive context. Using it within a *bounded* recursive context (e.g. only recurse when concurrently conflicting writes are detected) is very common and quite safe, but unbounded recursion is a no-no for the same reason that any non-tail-recursive unbounded recursion is problematic.

**Comment by armanbilge** [comment]
> This is totally valid and fine, but without the plugin it results in desugared code which will eventually go OOM.

That plugin or the equivalent functionality are not available for Scala 3. So we should probably rewrite the snippet.

**Comment by tomohavvk** [comment]
@djspiewak Thank you for the quick and detailed response. Now such behavior is clear to me.

It was my mistake; initially, I decided to visually familiarize myself with the tutorial, and the next day I moved on to practice. So I created a project and added the `cats-effect` dependency from memory, completely forgetting about `better-monadic-for`. 
Now I'll remember it for life :D


> We should probably make that more clear

Completely agree! 

Was surprised because I saw this declaration in the tutorial:
> as `IO` is stack safe we are not concerned about stack overflow issues

Although, as you explained above, these issues are not related in my case, it maybe worth rephrasing this and adding a warning about memory overflow in certain cases if the plugin is not used or just add a little chapter to [Starvation and Tuning](https://typelevel.org/cats-effect/docs/core/starvation-and-tuning)
